### ISSUE TITLE
Keepalived container broke when docker kill or reboot

### ISSUE THREAD

**Issue by winggundamth** [acassen/keepalived#383]
I'm using my own docker image **winggundamth/keepalived** https://github.com/winggundamth/docker-base/blob/trusty/keepalived/Dockerfile

Everything working fine if I do everything normally. But the problem will happen if I do command ```docker kill keepalived``` or try to force reboot host machine. Container will die and won't start again. I have to delete and create the new one.

This is Keepalived container information

```bash
# cat /etc/*release*
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=14.04
DISTRIB_CODENAME=trusty
DISTRIB_DESCRIPTION="Ubuntu 14.04.4 LTS"
NAME="Ubuntu"
VERSION="14.04.4 LTS, Trusty Tahr"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 14.04.4 LTS"
VERSION_ID="14.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"

# uname -a
Linux haproxy-test-1 4.4.0-31-generic #50~14.04.1-Ubuntu SMP Wed Jul 13 01:07:32 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

# keepalived --version
Keepalived v1.2.23 (07/11,2016)

Copyright (C) 2001-2016 Alexandre Cassen, <acassen@gmail.com>

Build options: KRNL_2_6 WITH_LVS HAVE_IPVS_SYNCD WITH_VRRP HAVE_VRRP_VMAC WITHOUT_ADDR_GEN_MODE WITHOUT_SNMP WITHOUT_SNMP_KEEPALIVED WITHOUT_SNMP_CHECKER WITHOUT_SNMP_RFC WITHOUT_SNMP_RFCV2 WITHOUT_SNMP_RFCV3 LIBIPVS_USE_NL WITHOUT_LIBNL WITH_VRRP_AUTH WITH_SO_MARK WITHOUT_LIBIPTC WITHOUT_LIBIPSET WITHOUT_IPV4_DEVCONF WITHOUT_IF_H_LINK_H_COLLISION HAVE_LINUX_NET_IF_H_COLLISION HAVE_SOCK_NONBLOCK HAVE_SOCK_CLOEXEC HAVE_FIB_ROUTING NO_MEM_CHECK NO_MEM_CHECK_LOG
```

This is when I try to do ```docker restart``` that working correctly

```bash
⟫ docker ps -a
CONTAINER ID        IMAGE                            COMMAND                  CREATED             STATUS              PORTS                                                                     NAMES
4f79e1f4ea1d        winggundamth/keepalived:trusty   "keepalived --use-fil"   5 minutes ago       Up 5 minutes                                                                                  keepalived
⟫ docker restart keepalived 
keepalived
ubuntu@haproxy-test-1:~⟫ docker ps -a
CONTAINER ID        IMAGE                            COMMAND                  CREATED             STATUS              PORTS                                                                     NAMES
4f79e1f4ea1d        winggundamth/keepalived:trusty   "keepalived --use-fil"   6 minutes ago       Up 1 seconds                                                                                  keepalived
⟫ docker logs keepalived 
Starting Healthcheck child process, pid=7
Starting VRRP child process, pid=8
Initializing ipvs 2.6
Registering Kernel netlink reflector
Registering Kernel netlink command channel
Registering Kernel netlink reflector
Registering Kernel netlink command channel
Registering gratuitous ARP shared channel
Opening file '/etc/keepalived/keepalived.conf'.
Opening file '/etc/keepalived/keepalived.conf'.
Using LinkWatch kernel netlink reflector...
Using LinkWatch kernel netlink reflector...
VRRP_Script(pingable_check_script) succeeded
VRRP_Instance(haproxy-test) Transition to MASTER STATE
VRRP_Instance(haproxy-test) Entering MASTER STATE
Stopping
Stopped
Stopped
Stopped Keepalived v1.2.23 (07/11,2016)
Starting Healthcheck child process, pid=7
Starting VRRP child process, pid=8
Initializing ipvs 2.6
Registering Kernel netlink reflector
Registering Kernel netlink reflector
Registering Kernel netlink command channel
Registering Kernel netlink command channel
Registering gratuitous ARP shared channel
Opening file '/etc/keepalived/keepalived.conf'.
Opening file '/etc/keepalived/keepalived.conf'.
Using LinkWatch kernel netlink reflector...
Using LinkWatch kernel netlink reflector...
VRRP_Script(pingable_check_script) succeeded
VRRP_Instance(haproxy-test) Transition to MASTER STATE
VRRP_Instance(haproxy-test) Entering MASTER STATE
```

This is when I try to do ```docker kill``` that causing problem. This will be the same if I force reboot host instantly

```bash
⟫ docker ps -a
CONTAINER ID        IMAGE                            COMMAND                  CREATED             STATUS              PORTS                                                                     NAMES
4f79e1f4ea1d        winggundamth/keepalived:trusty   "keepalived --use-fil"   7 minutes ago       Up About a minute                                                                             keepalived
ubuntu@haproxy-test-1:~⟫ docker kill keepalived 
keepalived
ubuntu@haproxy-test-1:~⟫ docker ps -a
CONTAINER ID        IMAGE                            COMMAND                  CREATED             STATUS                       PORTS                                                                     NAMES
4f79e1f4ea1d        winggundamth/keepalived:trusty   "keepalived --use-fil"   7 minutes ago       Exited (137) 3 seconds ago                                                                             keepalived
⟫ docker start keepalived 
keepalived
ubuntu@haproxy-test-1:~⟫ docker ps -a
CONTAINER ID        IMAGE                            COMMAND                  CREATED             STATUS                                  PORTS                                                                     NAMES
4f79e1f4ea1d        winggundamth/keepalived:trusty   "keepalived --use-fil"   8 minutes ago       Restarting (0) Less than a second ago                                                                             keepalived
ubuntu@haproxy-test-1:~⟫ docker logs keepalived 
Starting Healthcheck child process, pid=7
Starting VRRP child process, pid=8
Initializing ipvs 2.6
Registering Kernel netlink reflector
Registering Kernel netlink reflector
Registering Kernel netlink command channel
Registering Kernel netlink command channel
Registering gratuitous ARP shared channel
Opening file '/etc/keepalived/keepalived.conf'.
Opening file '/etc/keepalived/keepalived.conf'.
Using LinkWatch kernel netlink reflector...
Using LinkWatch kernel netlink reflector...
VRRP_Script(pingable_check_script) succeeded
VRRP_Instance(haproxy-test) Transition to MASTER STATE
VRRP_Instance(haproxy-test) Entering MASTER STATE
```

I'm not sure that this is expected or it won't support docker kill or force reboot so I know that I should use keepalived on host instead of container. Please let me know if you need more information

**Comment by pqarmitage** [comment]
If you do:
`docker kill keepalived`
`docker start keepalived`
`docker logs keepalived`

it will show at the end:
Keepalived[1]: Starting Keepalived v1.2.23 (07/11,2016)
Keepalived[1]: daemon is already running

The keepalived process terminates and since it is process 1, the container terminates. This would be expected behaviour.

**Comment by winggundamth** [comment]
Please see that when I'm doing ```docker start keepalived``` again. It won't start anymore and it doesn't show any log on ```docker logs``` command

```
ubuntu@haproxy-test-1:~⟫ docker start keepalived 
keepalived
ubuntu@haproxy-test-1:~⟫ docker ps -a
CONTAINER ID        IMAGE                            COMMAND                  CREATED             STATUS                                  PORTS                                                                     NAMES
4f79e1f4ea1d        winggundamth/keepalived:trusty   "keepalived --use-fil"   8 minutes ago       Restarting (0) Less than a second ago                                                                             keepalived
```

**Comment by pqarmitage** [comment]
@winggundamth you appear to have fewer entries in your logs that I do. For example, the following is what I see with `docker logs keepalived` after a normal start:
```
Keepalived[1]: Starting Keepalived v1.2.23 (07/11,2016)
Starting VRRP child process, pid=6
Keepalived[1]: Starting VRRP child process, pid=6
Registering Kernel netlink reflector
Keepalived_vrrp[6]: Registering Kernel netlink reflector
Registering Kernel netlink command channel
Keepalived_vrrp[6]: Registering Kernel netlink command channel
Registering gratuitous ARP shared channel
Keepalived_vrrp[6]: Registering gratuitous ARP shared channel
Opening file '/etc/keepalived/keepalived.conf'.
Keepalived_vrrp[6]: Opening file '/etc/keepalived/keepalived.conf'.
(VI_1): Cannot start in MASTER state if not address owner
Keepalived_vrrp[6]: (VI_1): Cannot start in MASTER state if not address owner
Using LinkWatch kernel netlink reflector...
Keepalived_vrrp[6]: Using LinkWatch kernel netlink reflector...
VRRP_Instance(VI_1) Entering BACKUP STATE
Keepalived_vrrp[6]: VRRP_Instance(VI_1) Entering BACKUP STATE
VRRP_Instance(VI_1) Transition to MASTER STATE
Keepalived_vrrp[6]: VRRP_Instance(VI_1) Transition to MASTER STATE
VRRP_Instance(VI_1) Entering MASTER STATE
Keepalived_vrrp[6]: VRRP_Instance(VI_1) Entering MASTER STATE
```
so every message you are getting, I am also getting prefixed by Keepalived...[PID], so it looks as though `docker logs` for you is only producing the output to the console, whereas for me it is giving both the console output and the syslog output. "Starting Keepalived" and "daemon is already running" messages are only output to the syslog.

The container I am running is built from the Dockerfile you referenced at the beginning of your initial message, so I'm not clear why the output of `dockers logs` should be different.

I've added a patch (ac7c717) included in pull request #384 which should log the starting and "already running" messages to the console when --log-console is specified, so you will then be able to see the log messages even if you are not getting the syslog messages.
